03_Modules/OcppRouter/src/module/webhook.dispatcher.ts

```diff
@@ -0,0 +1,288 @@
+// Copyright (c) 2023 S44, LLC
+// Copyright Contributors to the CitrineOS Project
+//
+// SPDX-License-Identifier: Apache 2.0
+
+import { MessageOrigin } from '@citrineos/base';
+import { ISubscriptionRepository, Subscription } from '@citrineos/data';
+import { ILogObj, Logger } from 'tslog';
+
+export class WebhookDispatcher {
+  private static readonly SUBSCRIPTION_REFRESH_INTERVAL_MS = 3 * 60 * 1000;
+
+  private _logger: Logger<ILogObj>;
+  private _subscriptionRepository: ISubscriptionRepository;
+
+  private _identifiers: Set<string> = new Set();
+
+  // Structure of the maps: key = identifier, value = array of callbacks
+  private _onConnectionCallbacks: Map<string, OnConnectionCallback[]> = new Map();
+  private _onCloseCallbacks: Map<string, OnCloseCallback[]> = new Map();
+  private _onMessageCallbacks: Map<string, OnMessageCallback[]> = new Map();
+  private _sentMessageCallbacks: Map<string, OnSentMessageCallback[]> = new Map();
+
+  constructor(
+    subscriptionRepository: ISubscriptionRepository,
+    logger?: Logger<ILogObj>,
+  ) {
+    this._subscriptionRepository = subscriptionRepository;
+    this._logger = logger
+      ? logger.getSubLogger({ name: this.constructor.name })
+      : new Logger<ILogObj>({ name: this.constructor.name });
+
+    setInterval(async () => {
+      await this._refreshSubscriptions();
+    }, WebhookDispatcher.SUBSCRIPTION_REFRESH_INTERVAL_MS);
+  }
+
+  addOnConnectionCallback(
+    identifier: string,
+    onConnectionCallback: OnConnectionCallback,
+  ): void {
+    this._onConnectionCallbacks.has(identifier)
+      ? this._onConnectionCallbacks.get(identifier)!.push(onConnectionCallback)
+      : this._onConnectionCallbacks.set(identifier, [onConnectionCallback]);
+  }
+
+  addOnCloseCallback(
+    identifier: string,
+    onCloseCallback: OnCloseCallback,
+  ): void {
+    this._onCloseCallbacks.has(identifier)
+      ? this._onCloseCallbacks.get(identifier)!.push(onCloseCallback)
+      : this._onCloseCallbacks.set(identifier, [onCloseCallback]);
+  }
+
+  addOnMessageCallback(
```

Comment:
I am a bit confused about the flow here. So when we add a subscription it's done via the  API which directly creates the data in the database. Then every once in a while you refresh the subscriptions by getting them from the database. But doesn't that introduce a long delay until the subscriptions get picked up? Shouldn't the `add...` be called when the subscription is added via API?

Reply:
Ah, apologies, I might have been over optimizing here. We had discussed that 3 min delay was acceptable against introducing a bunch of new complexity of getting instant consistency between the instances.

Reply:
@ChrisWeissmann just for posterity--after we move away from Directus, we will revisit this. the need for polling the database may go away after adopting a new UI stack, since it's data entry via Directus that this polling is needed to catch.